oauth2_http/java/com/google/auth/oauth2/JwtClaims.java

```diff
@@ -111,14 +118,111 @@ public boolean isComplete() {
 
   @AutoValue.Builder
   public abstract static class Builder {
+    /** Basic types supported by JSON standard. */
+    private static List<Class<? extends Serializable>> SUPPORTED_BASIC_TYPES = ImmutableList.of(
+            String.class, Integer.class, Double.class, Float.class, Boolean.class, Date.class,
+            String[].class, Integer[].class, Double[].class, Float[].class, Boolean[].class, Date[].class);
+
+    private static final String ERROR_MESSAGE = "Invalid type on additional claims. Valid types are String, Integer, " +
+            "Double, Float, Boolean, Date, List and Map. Map keys must be Strings.";
+
     public abstract Builder setAudience(String audience);
 
     public abstract Builder setIssuer(String issuer);
 
     public abstract Builder setSubject(String subject);
 
-    public abstract Builder setAdditionalClaims(Map<String, String> additionalClaims);
-
-    public abstract JwtClaims build();
+    public abstract Builder setAdditionalClaims(Map<String, ?> additionalClaims);
+
+    public abstract JwtClaims autoBuild();
+
+    public JwtClaims build() {
+      JwtClaims claims = autoBuild();
+      Preconditions.checkState(validateClaims(claims.getAdditionalClaims()), ERROR_MESSAGE);
+      return claims;
+    }
+
+    /**
+     * Validate if the objects on a Map are valid for a JWT claim.
+     *
+     * @param claims Map of claim objects to be validated
+     */
+    private static boolean validateClaims(Map<String, ?> claims) {
```

Comment:
This is interesting code. I wonder if there's a library for this somewhere?

Reply:
I asked a fellow programmer friends who are more experience than me on Java before coming up with this and they didn't know of any.

I ultimately got inspired by the logic used on [java-jwt](https://github.com/auth0/java-jwt/commit/343c3a2454f889c399ceed755bb5f6d33f2e2496?diff=unified) though I wanted to provide a more robust support than that implementation.

Open to any suggestions you may have on how to improve this code!